Homework 1

gfontenot opened the pull request:
[Assignment](http://www.seas.upenn.edu/~cis194/fall14/spring13/hw/01-intro.pdf)

gfontenot committed:
Add basic test setup using hspec

gfontenot committed:
implement `toDigits`

 - Turn the integer into a string with `show`
 - map `readChar` over the resulting string to transform the string from
   `[Char]` to `[Integer]`

gfontenot committed:
implement `toDigitsRev`

This just uses `toDigits`, and reverses the result

gfontenot committed:
Don't blow up on negative numbers

gfontenot committed:
Refactor tests

`toDigits` and `toDigitsRev` have some shared behavior. We can extract
that out into its own function for simplicity.

gfontenot committed:
Implement doubleEveryOther

This function doubles every other integer in a list starting from the
right. So it doubles the second to last element in a list and then goes
back from there.

The two `reverse` things are _dumb_ and I hate them, but I'm unclear on
how else to get this to work, and it's a good enough first pass.

gfontenot committed:
Write a worse version of `doubleEveryOther`

This is way worse, but works and doesn't use `reverse`.

gfontenot committed:
Go back to the good version of `doubleEveryOther`

I'm actually OK with this.

gabebw commented on spring-2013/hw1/src/CreditCardValidator.hs:
```diff
@@ -0,0 +1,15 @@
+module CreditCardValidator (toDigits, toDigitsRev, doubleEveryOther) where
+
+toDigits :: Integer -> [Integer]
+toDigits xs
```
What about renaming `xs` to `n`, since it's a number, not a list?

And:

``` haskell
toDigits n
  | n <= 0 = []
  | otherwise = map readChar $ show n
```

gfontenot replied:
uh, yes, why did I bail on the guard? It's a mystery to everyone involved.

gabebw commented on spring-2013/hw1/src/CreditCardValidator.hs:
```diff
@@ -0,0 +1,15 @@
+readChar = read . (:[])
+
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther = reverse . zipWith (*) (cycle [1, 2]) . reverse
```
Here's a cool version that I think Mike came up with: 

``` haskell
reverse . zipWith ($) (cycle [id, (*2)]) . reverse 
```

It's zipping with `($)`.

gfontenot replied:
oh, that's wild. I _think_ that zipping with `*` is clearer to me, but cool trick.

gabebw replied:
Yeah, I agree.

gabebw commented:
(Maybe you don't want feedback yet? In which case: oops, sorry.)

gfontenot commented:
@gabebw i always want feedback from you buddy

gfontenot committed:
Don't bail on the guard statement

why did i do this

gfontenot committed:
Implement sumDigits

This was _way_ easier than I was expecting. Plus: pointfree monads!

gfontenot committed:
Reorganize functions

Public functions up top, private down below

gfontenot committed:
Fix up style for module export

gfontenot committed:
Implement validate

Added `checkRemainder` instead of inlining the modulo stuff for
readability

gfontenot committed:
Add setup for Hanoi exercise

gfontenot committed:
Solve hanoi for 3 pegs

 - The one move we _know_ we're going to make is moving the disc from A
   to C (this is the solution for a single disc)
 - For each subsequent disc, we have to do two more moves: A -> B
   before, and then B -> C after. We can accomplish this by recursing
   into `hanoi` and passing B in the place of C for the previous step,
   and B instead of A for the subsequent step.
 - Somehow, because of math, this ends up doing the reverse steps that
   need to be performed as well.

I'm going to be honest: I barely understand how this actually works. But
I understand the code well enough, which is kind of the point.

gfontenot committed:
Remove toDigitsRev

This is a stupid function that wasn't even used in the final exercise.

gfontenot committed:
Move helper functions into where statements

These were only used in one place. I think I like it more where it's
obvious that they are intended to be helper functions instead of proper
top level functions.

gfontenot commented:
I think I'm happy with this now.

I wonder if Professors @pbrisbin and @gabebw would be so kind as to give me a review?

gabebw commented on spring-2013/hw1/src/CreditCardValidator.hs:
```diff
@@ -0,0 +1,22 @@
+doubleEveryOther = reverse . zipWith (*) (cycle [1, 2]) . reverse
+
+sumDigits :: [Integer] -> Integer
+sumDigits = sum . (toDigits =<<)
```
I'd probably write this as `sum . concatMap toDigits`.

Still: very fancy.

pbrisbin replied:
I agree avoiding the fanciness is better. As a rule, if my function is list-only, I use list-specific constructs. That's because the ratio of confusion/usefulness is not very good for the list monad (IMO).

gabebw commented on spring-2013/hw1/src/CreditCardValidator.hs:
```diff
@@ -0,0 +1,22 @@
+
+validate :: Integer -> Bool
+validate = checkRemainder . sumDigits . doubleEveryOther . toDigits
+    where checkRemainder = (== 0) . (`mod` 10)
```
How about naming this `isDivisibleBy10`?

gfontenot replied:
To be honest, I kinda like this named as an action. Reading this as `check remainder after summing the digits after doubling every other after converting to digits` feels right

gabebw replied:
Yeah, makes sense.

gabebw commented on spring-2013/hw1/test/CreditCardValidatorSpec.hs:
```diff
@@ -0,0 +1,31 @@
+    describe "doubleEveryOther" $ do
+        it "doubles every other digit in a list from the right" $ do
+            doubleEveryOther [1, 2, 3] `shouldBe` [1, 4, 3]
+            doubleEveryOther [8,7,6,5] `shouldBe` [16,7,12,5]
```
spaces after commas?

gabebw commented on spring-2013/hw1/test/CreditCardValidatorSpec.hs:
```diff
@@ -0,0 +1,31 @@
+
+    describe "sumDigits" $ do
+        it "returns the sum of every digit in the list" $
+            sumDigits [16,7,12,5] `shouldBe` 22
```
spaces, commas, you know the deal

gabebw commented on spring-2013/hw1/src/CreditCardValidator.hs:
```diff
@@ -0,0 +1,22 @@
+toDigits :: Integer -> [Integer]
+toDigits n
+    | n <= 0 = []
+    | otherwise = map readChar . show $ n
```
I think I prefer `map readChar $ show n`

gfontenot replied:
yeah, good call. I think this is a holdover from the fact that this _used_ to be a pointfree function before I needed to check the value of `n` which screwed everything up.

gfontenot committed:
Add spaces after commas

Copying and pasting code is bad and you shouldn't do it

gfontenot committed:
Improve style for toDigits

This used to be a pointfree function, which was super nice. Then I had
to check the value of `n` to make sure if was positive, and that screwed
everything up. To get it to compile, I just tossed an `$` at the end,
but it actually reads better if we use function application _instead_ of
function composition.

gabebw commented on spring-2013/hw1/src/CreditCardValidator.hs:
```diff
@@ -0,0 +1,22 @@
+toDigits n
+    | n <= 0 = []
+    | otherwise = map readChar $ show n
+    where readChar = read . (:[])
```
Interesting, I always put helper functions on their own lines:

```
where
    readChar = read . (:[])
```

I think it makes it easier to see that there's a helper at all.

pbrisbin replied:
I also line-break always, though I prefer the de-dented `where`

``` hs
toDigits n
    | ...
    | otherwise ...
  where
    readChar = ...
```

That said, swooping up single-line sections into a tighter phrase can work well sometimes:

``` hs
strip = trim . trim where trim = reverse . dropWhile isSpace
```

gfontenot replied:
Yep, I like that much more. Thanks!

gfontenot committed:
Update `where` style for readability

gfontenot committed:
Use concatMap instead of bind

@gabebw and @pbrisbin hate cool things
